File: packages/arabic_bidi/dart/test/bidi_test.dart
import 'package:test/test.dart';
import 'package:arabic_bidi/arabic_bidi.dart';

// Expected values are written as Unicode escape sequences so the intent is
// unambiguous regardless of editor font/direction.
//
// Quick form-index reference per Presentation Forms-B layout:
//   [iso, fin, ini, med]  ← index order in the _forms map
//
// Selected letter lookups:
//   ء  0x0621 → iso=FE80  (U-type, no connecting forms)
//   آ  0x0622 → iso=FE81  fin=FE82
//   أ  0x0623 → iso=FE83  fin=FE84
//   ؤ  0x0624 → iso=FE85  fin=FE86  (R-type)
//   إ  0x0625 → iso=FE87  fin=FE88
//   ا  0x0627 → iso=FE8D  fin=FE8E  (R-type)
//   ب  0x0628 → iso=FE8F  fin=FE90  ini=FE91  med=FE92
//   ت  0x062A → iso=FE95  fin=FE96  ini=FE97  med=FE98
//   ح  0x062D → iso=FEA1  fin=FEA2  ini=FEA3  med=FEA4
//   د  0x062F → iso=FEA9  fin=FEAA  (R-type)
//   ذ  0x0630 → iso=FEAB  fin=FEAC  (R-type)
//   ر  0x0631 → iso=FEAD  fin=FEAE  (R-type)
//   ز  0x0632 → iso=FEAF  fin=FEB0  (R-type)
//   س  0x0633 → iso=FEB1  fin=FEB2  ini=FEB3  med=FEB4
//   ش  0x0634 → iso=FEB5  fin=FEB6  ini=FEB7  med=FEB8
//   ف  0x0641 → iso=FED1  fin=FED2  ini=FED3  med=FED4
//   ق  0x0642 → iso=FED5  fin=FED6  ini=FED7  med=FED8
//   ك  0x0643 → iso=FED9  fin=FEDA  ini=FEDB  med=FEDC
//   ل  0x0644 → iso=FEDD  fin=FEDE  ini=FEDF  med=FEE0
//   م  0x0645 → iso=FEE1  fin=FEE2  ini=FEE3  med=FEE4
//   ن  0x0646 → iso=FEE5  fin=FEE6  ini=FEE7  med=FEE8
//   ه  0x0647 → iso=FEE9  fin=FEEA  ini=FEEB  med=FEEC
//   و  0x0648 → iso=FEED  fin=FEEE  (R-type)
//   ي  0x064A → iso=FEF1  fin=FEF2  ini=FEF3  med=FEF4

void main() {
  // ── Isolated forms ────────────────────────────────────────────────────────

  group('reshape — isolated', () {
    test('empty string', () => expect(ArabicBidi.reshape(''), equals('')));

    test('single letter م: isolated', () {
      // م (D-type): no neighbours → isolated = FEE1
      expect(ArabicBidi.reshape('م'), equals('ﻡ'));
    });

    test('ء: always isolated (U-type, no connecting forms)', () {
      expect(ArabicBidi.reshape('ء'), equals('ﺀ'));
    });

    test('اد: two R-type letters — both isolated', () {
      // ا(R) cannot extend to د; د preceding is ا(R, joinsRight=F) → no chain
      // ا iso=FE8D, د iso=FEA9
      expect(ArabicBidi.reshape('اد'), equals('ﺍﺩ'));
    });

    test('ورد (rose): all R-type — all isolated', () {
      // و iso=FEED, ر iso=FEAD, د iso=FEA9
      expect(ArabicBidi.reshape('ورد'), equals('ﻭﺭﺩ'));
    });

    test('دار (house): all R-type — all isolated', () {
      // د iso=FEA9, ا iso=FE8D, ر iso=FEAD
      expect(ArabicBidi.reshape('دار'), equals('ﺩﺍﺭ'));
    });

    test('آدم (Adam): R+R+D where D has no prev connector', () {
      // آ iso=FE81, د iso=FEA9, م iso=FEE1
      // م: prev=د(joinsRight=F) → isolated
      expect(ArabicBidi.reshape('آدم'), equals('ﺁﺩﻡ'));
    });

    test('رؤوف (Raouf): R+R+R+D where all break the chain', () {
      // ر iso=FEAD, ؤ iso=FE85, و iso=FEED, ف iso=FED1
      // ف: prev=و(joinsRight=F) → isolated
      expect(ArabicBidi.reshape('رؤوف'), equals('ﺭﺅﻭﻑ'));
    });
  });

  // ── Final forms ───────────────────────────────────────────────────────────

  group('reshape — final', () {
    test('بر: D + R → initial + final', () {
      // ب ini=FE91, ر fin=FEAE
      expect(ArabicBidi.reshape('بر'), equals('ﺑﺮ'));
    });

    test('بت: D + D (no next) → initial + final', () {
      // ب ini=FE91, ت fin=FE96
      expect(ArabicBidi.reshape('بت'), equals('ﺑﺖ'));
    });

    test('زيد (Zayd): R + D + R → iso + ini + fin', () {
      // ز(R) iso=FEAF; ي(D) prev=ز(JR=F) → ini=FEF3; د(R) prev=ي(JR=T) → fin=FEAA
      expect(ArabicBidi.reshape('زيد'), equals('ﺯﻳﺪ'));
    });
  });

  // ── Medial forms ──────────────────────────────────────────────────────────

  group('reshape — medial', () {
    test('بيت (house): D + D + D → ini + med + fin', () {
      // ب ini=FE91, ي med=FEF4, ت fin=FE96
      expect(ArabicBidi.reshape('بيت'), equals('ﺑﻴﺖ'));
    });

    test('بيب: D + D + D → ini + med + fin', () {
      // ب ini=FE91, ي med=FEF4, ب fin=FE90
      expect(ArabicBidi.reshape('بيب'), equals('ﺑﻴﺐ'));
    });

    test('محمد (Muhammad): ini + med + med + fin', () {
      // م ini=FEE3, ح med=FEA4, م med=FEE4, د fin=FEAA
      // Third م: prev=ح(JR=T), next=د(JL=T) → medial (د is R-type but joinsLeft=T)
      expect(ArabicBidi.reshape('محمد'), equals('ﻣﺤﻤﺪ'));
    });

    test('سلام (peace): ini + med + fin + iso', () {
      // س ini=FEB3, ل med=FEE0 (next=ا JL=T), ا fin=FE8E, م iso=FEE1
      // م: prev=ا(JR=F) → isolated
      expect(ArabicBidi.reshape('سلام'), equals('ﺳﻠﺎﻡ'));
    });

    test('بنت (girl): ini + med + fin', () {
      // ب ini=FE91, ن med=FEE8, ت fin=FE96
      expect(ArabicBidi.reshape('بنت'), equals('ﺑﻨﺖ'));
    });
  });

  // ── Chain breaks (R-type in the middle) ───────────────────────────────────

  group('reshape — chain breaks', () {
    test('باب (door): ini + fin + iso', () {
      // ب ini=FE91; ا fin=FE8E; ب iso=FE8F (prev=ا JR=F → no chain)
      expect(ArabicBidi.reshape('باب'), equals('ﺑﺎﺏ'));
    });

    test('مرحبا: chain broken by ر', () {
      // م ini=FEE3; ر fin=FEAE (prev=م JR=T); ح ini=FEA3 (new chain after ر);
      // ب med=FE92; ا fin=FE8E
      expect(ArabicBidi.reshape('مرحبا'), equals('ﻣﺮﺣﺒﺎ'));
    });

    test('أمل (hope): R + D + D → iso + ini + fin', () {
      // أ iso=FE83; م ini=FEE3 (prev=أ JR=F → new chain); ل fin=FEDE
      expect(ArabicBidi.reshape('أمل'), equals('ﺃﻣﻞ'));
    });

    test('مسؤول: chain broken by ؤ and و', () {
      // م ini=FEE3, س med=FEB4, ؤ fin=FE86 (prev=س JR=T; ؤ R-type → fin)
      // و iso=FEED (prev=ؤ JR=F), ل iso=FEDD (prev=و JR=F)
      expect(ArabicBidi.reshape('مسؤول'), equals('ﻣﺴﺆﻭﻝ'));
    });
  });

  // ── ء (U-type) halts connections ──────────────────────────────────────────

  group('reshape — hamza and hamza-containing words', () {
    test('ء: isolated', () => expect(ArabicBidi.reshape('ء'), equals('ﺀ')));

    test('شيء (thing): D + D + U → ini + fin + iso', () {
      // ش ini=FEB7; ي: prev=ش(JR=T), next=ء(JL=F U-type) → fin=FEF2; ء iso=FE80
      expect(ArabicBidi.reshape('شيء'), equals('ﺷﻲﺀ'));
    });

    test('ماء (water): D + R + U → ini + fin + iso', () {
      // م ini=FEE3; ا fin=FE8E (prev=م JR=T); ء iso=FE80
      expect(ArabicBidi.reshape('ماء'), equals('ﻣﺎﺀ'));
    });
  });

  // ── Tashkeel and tatweel transparency ────────────────────────────────────

  group('reshape — transparent characters', () {
    test('tashkeel marks are preserved and transparent for joining', () {
      // مُحَمَّدٌ → same reshaping as محمد; tashkeel interspersed
      // م FEE3 + U+064F + ح FEA4 + U+064E + م FEE4 + U+064E + U+0651 + د FEAA
      expect(
        ArabicBidi.reshape('مُحَمَّدٌ'),
        equals('ﻣُﺤَﻤَّﺪٌ'),
      );
    });

    test('مُحَمَّد (without trailing tanwin): same reshaping as محمد', () {
      expect(
        ArabicBidi.reshape('مُحَمَّد'),
        equals('ﻣُﺤَﻤَّﺪ'),
      );
    });

    test('مـحـمـد (tatweels transparent): same letter forms as محمد', () {
      // Tatweels (U+0640) are preserved in output but skipped for joining context
      expect(
        ArabicBidi.reshape('مـحـمـد'),
        equals('ﻣـﺤـﻤـﺪ'),
      );
    });

    test('بـت (tatweel between D-type letters): ini + transparent + fin', () {
      // ب ini=FE91; U+0640 pass-through; ت fin=FE96
      expect(ArabicBidi.reshape('بـت'), equals('ﺑـﺖ'));
    });
  });

  // ── Non-Arabic passthrough ────────────────────────────────────────────────

  group('reshape — non-Arabic passthrough', () {
    test('empty string', () => expect(ArabicBidi.reshape(''), equals('')));

    test('pure Latin unchanged', () {
      expect(ArabicBidi.reshape('Hello World'), equals('Hello World'));
    });

    test('Arabic+Latin: only Arabic letters are reshaped', () {
      // مرحبا: م ini=FEE3, ر fin=FEAE, ح ini=FEA3, ب med=FE92, ا fin=FE8E
      expect(
        ArabicBidi.reshape('Hello مرحبا World'),
        equals('Hello ﻣﺮﺣﺒﺎ World'),
      );
    });

    test('كود Python: Arabic word followed by Latin', () {
      // ك ini=FEDB, و fin=FEEE, د iso=FEA9 (prev=و JR=F)
      expect(ArabicBidi.reshape('كود Python'), equals('ﻛﻮﺩ Python'));
    });

    test('رقم 123 هاتف: digits unchanged; Arabic words shaped independently',
        () {
      // رقم: ر iso=FEAD, ق ini=FED7, م fin=FEE2
      // هاتف: ه ini=FEEB, ا fin=FE8E, ت ini=FE97, ف fin=FED2
      expect(
        ArabicBidi.reshape('رقم 123 هاتف'),
        equals('ﺭﻗﻢ 123 ﻫﺎﺗﻒ'),
      );
    });

    test('digits pass through unchanged', () {
      expect(ArabicBidi.reshape('السعر ١٢٣'), contains('١٢٣'));
    });

    test('space between words restarts joining context', () {
      // من: م ini=FEE3, ن fin=FEE6
      // نحن: ن ini=FEE7, ح med=FEA4, ن fin=FEE6
      expect(ArabicBidi.reshape('من نحن'), equals('ﻣﻦ ﻧﺤﻦ'));
    });
  });

  // ── Full word suite from task spec ───────────────────────────────────────

  group('reshape — task word suite', () {
    test('لا (no): ل + ا → ini + fin (no ligature in default mode)', () {
      // ل ini=FEDF, ا fin=FE8E
      expect(ArabicBidi.reshape('لا'), equals('ﻟﺎ'));
    });

    test('السلام (as-salamu): ا + ل + س + ل + ا + م', () {
      // ا iso=FE8D; ل ini=FEDF; س med=FEB4; ل med=FEE0; ا fin=FE8E; م iso=FEE1
      expect(ArabicBidi.reshape('السلام'), equals('ﺍﻟﺴﻠﺎﻡ'));
    });
  });

  // ── Lam-alef ligatures (opt-in) ───────────────────────────────────────────

  group('reshape — lam-alef ligatures disabled (default)', () {
    test('لا: default mode → initial ل + final ا (no ligature)', () {
      expect(ArabicBidi.reshape('لا'), equals('ﻟﺎ'));
    });

    test('كلا: default mode → ini ك + med ل + fin ا', () {
      expect(ArabicBidi.reshape('كلا'), equals('ﻛﻠﺎ'));
    });
  });

  group('reshape — lam-alef ligatures enabled', () {
    const opts = ArabicReshapeOptions(useLamAlefLigatures: true);

    test('لا → isolated lam-alef ligature U+FEFB', () {
      expect(ArabicBidi.reshape('لا', options: opts), equals('ﻻ'));
    });

    test('لأ → isolated lam-alef-hamza-above U+FEF7', () {
      expect(ArabicBidi.reshape('لأ', options: opts), equals('ﻷ'));
    });

    test('لإ → isolated lam-alef-hamza-below U+FEF9', () {
      expect(ArabicBidi.reshape('لإ', options: opts), equals('ﻹ'));
    });

    test('لآ → isolated lam-alef-madda U+FEF5', () {
      expect(ArabicBidi.reshape('لآ', options: opts), equals('ﻵ'));
    });

    test('كلا → final ligature U+FEFC (lam preceded by connector)', () {
      // ك ini=FEDB; ل preceded by ك (joinsRight=T) → FEFC (final lam-alef)
      expect(ArabicBidi.reshape('كلا', options: opts), equals('ﻛﻼ'));
    });

    test('السلام: second لا → final ligature U+FEFC', () {
      // ا iso=FE8D; ل ini=FEDF (next=س, not alef); س med=FEB4;
      // ل + ا → FEFC (prev=س JR=T → final ligature); م iso=FEE1
      expect(
        ArabicBidi.reshape('السلام', options: opts),
        equals('ﺍﻟﺴﻼﻡ'),
      );
    });

    test('الا: first لا → isolated ligature (prev ا is R-type)', () {
      // ا iso=FE8D; ل ini (prev=ا JR=F) + ا → FEFB (isolated); no more chars
      expect(ArabicBidi.reshape('الا', options: opts), equals('ﺍﻻ'));
    });

    test('ligature option passed via ArabicBidi.reshape', () {
      expect(
        ArabicBidi.reshape('لا',
            options: const ArabicReshapeOptions(useLamAlefLigatures: true)),
        equals('ﻻ'),
      );
    });
  });

  // ── ArabicBidi facade ─────────────────────────────────────────────────────

  group('ArabicBidi facade', () {
    test('ArabicBidi.reshape returns correct shaped forms', () {
      expect(ArabicBidi.reshape('بيت'), equals('ﺑﻴﺖ'));
    });

    test('explicit ArabicReshapeOptions.defaults matches implicit defaults',
        () {
      expect(
        ArabicBidi.reshape('لا', options: ArabicReshapeOptions.defaults),
        equals(ArabicBidi.reshape('لا')),
      );
    });
  });

  // ── Direction detection ───────────────────────────────────────────────────

  group('detectDirection', () {
    test('pure Arabic → rtl', () {
      expect(ArabicBidi.detectDirection('مرحبا'), equals(Direction.rtl));
    });

    test('pure Latin → ltr', () {
      expect(ArabicBidi.detectDirection('Hello'), equals(Direction.ltr));
    });

    test('mixed Arabic+Latin → mixed', () {
      expect(
        ArabicBidi.detectDirection('Hello مرحبا World'),
        equals(Direction.mixed),
      );
    });

    test('empty string → ltr', () {
      expect(ArabicBidi.detectDirection(''), equals(Direction.ltr));
    });
  });
}

File: packages/arabic_bidi/js/test/bidi.test.ts
import { describe, it, expect, vi, afterEach } from 'vitest';
import {
  reshape,
  detectDirection, Direction,
  prepareForTerminal,
  printArabic,
  ArabicLogger, arabicLogger,
} from '../src/index';

// Helpers — named constants for presentation form code points
const fcp = (...cps: number[]) => String.fromCodePoint(...cps);

// Isolated forms
const BA_ISO  = 0xFE8F, BA_INI  = 0xFE91, BA_MED  = 0xFE92, BA_FIN  = 0xFE90;
const NUN_FIN = 0xFEE6, NUN_MED = 0xFEE8;
const TA_FIN  = 0xFE96;
const ALEF_ISO = 0xFE8D, ALEF_FIN = 0xFE8E;
const RA_ISO   = 0xFEAD;
const WAW_ISO  = 0xFEED;
const YA_MED   = 0xFEF4, YA_INI  = 0xFEF3;
const LAM_INI  = 0xFEDF;
const RA_ISO_CP = 0xFEAD, QAF_INI = 0xFED7, MEEM_FIN = 0xFEE2;
const HAMZA_ISO = 0xFE80;
const LAM_ALEF_ISO = 0xFEFB, LAM_ALEF_FIN = 0xFEFC;
const LAM_MADDA_ISO = 0xFEF5;

// ── reshape ───────────────────────────────────────────────────────────────────

describe('reshape - basic', () => {
  it('empty string returns empty', () => {
    expect(reshape('')).toBe('');
  });

  it('Latin passes through unchanged', () => {
    expect(reshape('Hello World')).toBe('Hello World');
  });

  it('ASCII digits pass through unchanged', () => {
    expect(reshape('123')).toBe('123');
  });
});

describe('reshape - isolated forms', () => {
  it('hamza (U-type) -> isolated form U+FE80', () => {
    // hamza has no joining; always emits isolated form
    expect(reshape('ء')).toBe(fcp(HAMZA_ISO));
  });

  it('ba alone -> isolated form U+FE8F', () => {
    expect(reshape('ب')).toBe(fcp(BA_ISO));
  });
});

describe('reshape - initial form', () => {
  // ba(D) + nun(D): ba has no prev, next=nun joinsLeft -> initial; nun has prev=ba joinsRight, next=none -> final
  it('ba+nun: initial+final', () => {
    expect(reshape('بن')).toBe(fcp(BA_INI, NUN_FIN));
  });
});

describe('reshape - medial form', () => {
  // ba(D)+nun(D)+ta(D): ba->initial, nun->medial (both neighbours join), ta->final
  it('ba+nun+ta: initial+medial+final', () => {
    expect(reshape('بنت')).toBe(fcp(BA_INI, NUN_MED, TA_FIN));
  });
});

describe('reshape - R-type letters break chain', () => {
  // ba(D)+alef(R)+ra(R):
  //   ba -> initial (next=alef joinsLeft=true)
  //   alef -> final (prev=ba joinsRight=true; alef joinsRight=false so it cannot extend)
  //   ra -> isolated (prev=alef joinsRight=false, ra joinsRight=false -> not initial)
  it('ba+alef+ra: R-type letters do not extend the join rightward', () => {
    expect(reshape('بار')).toBe(fcp(BA_INI, ALEF_FIN, RA_ISO));
  });

  // waw(R)+alef(R): neither has a D-type predecessor to make it final/initial
  it('waw+alef: both R-type -> both isolated', () => {
    expect(reshape('وا')).toBe(fcp(WAW_ISO, ALEF_ISO));
  });
});

describe('reshape - transparent characters', () => {
  // ba + fatha(U+064E) + ya + ta: fatha is tashkeel, transparent
  // ba sees ya as next -> initial; ya sees ba as prev -> medial; ta -> final
  it('fatha (tashkeel U+064E) is transparent for joining, preserved in output', () => {
    expect(reshape('بَيت')).toBe(fcp(BA_INI, 0x064E, YA_MED, TA_FIN));
  });

  // ba + tatweel(U+0640) + ya + ta: same but tatweel
  it('tatweel (U+0640) is transparent for joining, preserved in output', () => {
    expect(reshape('بـيت')).toBe(fcp(BA_INI, 0x0640, YA_MED, TA_FIN));
  });
});

describe('reshape - lam-alef ligatures', () => {
  // lam+alef without option: lam->initial (next=alef joinsLeft), alef->final
  it('lam+alef default: no ligature, normal contextual forms', () => {
    expect(reshape('لا')).toBe(fcp(LAM_INI, ALEF_FIN));
  });

  // lam+alef with opt-in: isolated lam-alef ligature (U+FEFB)
  it('lam+alef with useLamAlefLigatures: isolated ligature U+FEFB', () => {
    expect(reshape('لا', { useLamAlefLigatures: true })).toBe(fcp(LAM_ALEF_ISO));
  });

  // ba+lam+alef: ba->initial; lam has prev=ba(joinsRight) -> use final lam-alef ligature (U+FEFC)
  it('ba+lam+alef with useLamAlefLigatures: ba-initial + final ligature U+FEFC', () => {
    expect(reshape('بلا', { useLamAlefLigatures: true })).toBe(fcp(BA_INI, LAM_ALEF_FIN));
  });

  // lam+alef-madda: isolated lam+alef-madda ligature (U+FEF5)
  it('lam+alef-madda with useLamAlefLigatures: isolated ligature U+FEF5', () => {
    expect(reshape('لآ', { useLamAlefLigatures: true })).toBe(fcp(LAM_MADDA_ISO));
  });
});

// ── detectDirection ───────────────────────────────────────────────────────────

describe('detectDirection', () => {
  it('empty string -> LTR', () => {
    expect(detectDirection('')).toBe(Direction.LTR);
  });

  it('pure Arabic text -> RTL (Arabic-block >= 80% of arabic+latin)', () => {
    // all 5 chars are Arabic-block, 0 Latin -> 100% -> RTL
    expect(detectDirection('مرحبا')).toBe(Direction.RTL);
  });

  it('pure Latin text -> LTR (Latin >= 80% of arabic+latin)', () => {
    expect(detectDirection('Hello')).toBe(Direction.LTR);
  });

  it('digits-only -> LTR (digits not counted as Arabic or Latin)', () => {
    expect(detectDirection('12345')).toBe(Direction.LTR);
  });

  it('equal Arabic and Latin -> MIXED (neither >= 80%)', () => {
    // 'Hello مرحبا': 5 Latin + 5 Arabic-block
    expect(detectDirection('Hello مرحبا')).toBe(Direction.MIXED);
  });

  it('50/50 split -> MIXED', () => {
    // 'ab مر': 2 Latin, 2 Arabic-block
    expect(detectDirection('ab مر')).toBe(Direction.MIXED);
  });

  it('Arabic >= 80% with small Latin minority -> RTL', () => {
    // 'السلام a': 6 Arabic-block + 1 Latin -> 6/7 >= 0.8 -> RTL
    expect(detectDirection('السلام a')).toBe(Direction.RTL);
  });
});

// ── prepareForTerminal ────────────────────────────────────────────────────────

describe('prepareForTerminal - Arabic-only text', () => {
  it('empty string -> empty', () => {
    expect(prepareForTerminal('')).toBe('');
  });

  it('pure Arabic: reshaped correctly (single run, reorder has no effect)', () => {
    // 'بيت' = ba+ya+ta (house)
    // arabicRatio = 1.0 > 0.5 -> reorder (but single run stays same)
    // reshape: ba-ini(FE91) + ya-med(FEF4) + ta-fin(FE96)
    expect(prepareForTerminal('بيت')).toBe(fcp(0xFE91, 0xFEF4, 0xFE96));
  });
});

describe('prepareForTerminal - mixed Arabic+Latin', () => {
  it('Arabic-dominant text: Latin run moves to front after reorder', () => {
    // 'السلام عليكم Hello'
    // 11 Arabic-block / 18 total = 0.61 > 0.5 -> reorder
    // reversed runs: Latin('Hello') + space + Arabic(reshaped) + space + Arabic(reshaped)
    const result = prepareForTerminal('السلام عليكم Hello');
    expect(result.startsWith('Hello')).toBe(true);
    // Arabic presentation-form chars present (U+FE70-U+FEFF range)
    expect(/[ﹰ-﻿]/.test(result)).toBe(true);
  });

  it('Latin-dominant text: run order unchanged', () => {
    // 'Hello ب': 1 Arabic-block / 8 total = 0.125 -> no reorder
    const result = prepareForTerminal('Hello ب');
    expect(result.startsWith('Hello')).toBe(true);
  });
});

describe('prepareForTerminal - digits and arabicLetterRatio', () => {
  it('Eastern Arabic digits NOT counted as letters: ratio 3/6=0.5 -> no reorder', () => {
    // 'رقم ١٢٣': 3 Arabic letters, 3 Eastern Arabic digits (not letters), 1 space
    // arabicLetterRatio = 3 letters / 6 non-ws = 0.5 — NOT > 0.5 -> no reorder
    // reshape 'رقم': ra-iso(FEAD) + qaf-ini(FED7) + meem-fin(FEE2)
    // result: reshaped-raqm + space + digits (original order)
    const expected = fcp(RA_ISO_CP, QAF_INI, MEEM_FIN, 0x20, 0x0661, 0x0662, 0x0663);
    expect(prepareForTerminal('رقم ١٢٣')).toBe(expected);
  });

  it('Persian digits NOT counted as letters: no reorder', () => {
    // 'رقم ۱۲۳': same structure with Persian digits U+06F1–U+06F3
    const result = prepareForTerminal('رقم ۱۲۳');
    // word comes first (no reorder), digits at end
    expect(result.endsWith('۱۲۳')).toBe(true);
  });

  it('ASCII digits are NumberRun: not counted as letters, no reorder', () => {
    // 'رقم 123': arabicLetterRatio = 3/6 = 0.5 -> no reorder
    const result = prepareForTerminal('رقم 123');
    expect(result.endsWith('123')).toBe(true);
  });

  it('Arabic-dominant text with ratio > 0.5 still reorders', () => {
    // 'السلام عليكم': 11 letters / 11 non-ws = 1.0 > 0.5 -> reorder
    // Result starts with last run reversed to front
    const result = prepareForTerminal('السلام Hello');
    expect(result.startsWith('Hello')).toBe(true);
  });

  it('50/50 text (ratio = 0.5) does NOT reorder', () => {
    // 3 Arabic letters + 3 Latin letters, no spaces: ratio = 3/6 = 0.5 -> no reorder
    // 'بنت' (3 Arabic) + 'abc' (3 Latin): first run is Arabic
    const result = prepareForTerminal('بنتabc');
    // Arabic run stays first (not reversed)
    expect(result.endsWith('abc')).toBe(true);
  });
});

describe('prepareForTerminal - options', () => {
  it('reshape: false -> Arabic not converted to presentation forms', () => {
    expect(prepareForTerminal('بيت', { reshape: false })).toBe('بيت');
  });

  it('reorder: false -> run order unchanged for Arabic-dominant text', () => {
    // Arabic still gets reshaped, but runs stay in original order
    const result = prepareForTerminal('السلام Hello', { reorder: false });
    // Arabic run comes first (not moved to end)
    expect(/^[ﹰ-﻿]/.test(result)).toBe(true);
  });
});

// ── ArabicLogger ──────────────────────────────────────────────────────────────

describe('ArabicLogger', () => {
  afterEach(() => vi.restoreAllMocks());

  it('info: [INFO] level label', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    arabicLogger.info('Hello');
    expect(spy).toHaveBeenCalledWith('[INFO] Hello');
  });

  it('error: [ERROR] level label', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    arabicLogger.error('boom');
    expect(spy).toHaveBeenCalledWith('[ERROR] boom');
  });

  it('warn: [WARN] level label', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    arabicLogger.warn('caution');
    expect(spy).toHaveBeenCalledWith('[WARN] caution');
  });

  it('debug: [DEBUG] level label', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    arabicLogger.debug('trace');
    expect(spy).toHaveBeenCalledWith('[DEBUG] trace');
  });

  it('prefix: [prefix][LEVEL] message', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    new ArabicLogger({ prefix: 'app' }).info('started');
    expect(spy).toHaveBeenCalledWith('[app][INFO] started');
  });

  it('useTimestamp: output starts with [YYYY-MM-DD HH:MM:SS]', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    new ArabicLogger({ useTimestamp: true }).info('x');
    expect(spy.mock.calls[0][0]).toMatch(/^\[20\d{2}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\]/);
  });

  it('prefix + timestamp: timestamp then prefix then level', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    new ArabicLogger({ prefix: 'srv', useTimestamp: true }).warn('down');
    expect(spy.mock.calls[0][0]).toMatch(/^\[20\d{2}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\]\[srv\]\[WARN\] down$/);
  });

  it('Arabic message is reshaped before output', () => {
    // ba+ya+ta reshaped: BA_INI(FE91) + YA_MED(FEF4) + TA_FIN(FE96)
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    arabicLogger.info('بيت');
    expect(spy).toHaveBeenCalledWith('[INFO] ' + fcp(0xFE91, 0xFEF4, 0xFE96));
  });

  it('printArabic calls console.log with prepared text', () => {
    const spy = vi.spyOn(console, 'log').mockImplementation(() => {});
    printArabic('بيت');
    expect(spy).toHaveBeenCalledWith(fcp(0xFE91, 0xFEF4, 0xFE96));
  });
});
